docs/_snippets/framework/tutorials/external-widget.js

```diff
@@ -0,0 +1,208 @@
+import Heading from '@ckeditor/ckeditor5-heading/src/heading';
+import List from '@ckeditor/ckeditor5-list/src/list';
+import Bold from '@ckeditor/ckeditor5-basic-styles/src/bold';
+import Italic from '@ckeditor/ckeditor5-basic-styles/src/italic';
+
+class ExternalWidgetCommand extends Command {
+	execute() {
+		const editor = this.editor;
+
+		const RESOURCE_URL = 'https://api2.binance.com/api/v3/ticker/24hr?symbol=BTCUSDT';
+
+		editor.model.change( writer => {
+			const externalWidget = writer.createElement(
+				'externalElement',
+				{
+					'data-resource-url': RESOURCE_URL
+				}
+			);
+
+			editor.model.insertContent( externalWidget );
+
+			writer.setSelection( externalWidget, 'on' );
+		} );
+	}
+}
+
+class ExternalWidget extends Plugin {
+	static get requires() {
+		return [ ExternalWidgetEditing, ExternalWidgetUI ];
+	}
+}
+
+class ExternalWidgetUI extends Plugin {
+	init() {
+		const editor = this.editor;
+
+		editor.ui.componentFactory.add( 'external', locale => {
+			const button = new ButtonView( locale );
+
+			button.set( {
+				label: 'Insert Bitcoin rate',
+				tooltip: true,
+				withText: true,
+				isEnabled: true
+			} );
+
+			button.on( 'execute', () => {
+				editor.execute( 'external' );
+				editor.editing.view.focus();
+			} );
+
+			return button;
+		} );
+	}
+}
+
+class ExternalWidgetEditing extends Plugin {
+	constructor( editor ) {
+		super( editor );
+
+		this._updateWidgetDataFn = this._updateWidgetData;
+
+		this.intervalId = this._intervalFetch();
+	}
+
+	static get requires() {
+		return [ Widget ];
+	}
+
+	destroy() {
+		clearInterval( this.intervalId );
+	}
+
+	init() {
+		const schema = this.editor.model.schema;
+
+		schema.register( 'externalElement', {
+			allowWhere: '$text',
+			isObject: true,
+			isInline: true,
+			allowChildren: '$text',
+			allowAttributesOf: '$text',
+			allowAttributes: [ 'data-resource-url' ]
+		} );
+
+		this._defineSchema();
+		this._defineConverters();
+
+		this.editor.commands.add( 'external', new ExternalWidgetCommand( this.editor ) );
+	}
+
+	_intervalFetch() {
+		return setInterval( this._updateWidgetDataFn.bind( this ), 20000 ); // set time interval to 20s
+	}
+
+	async _updateWidgetData(
+		insertFetch = false,
+		externalUrl = 'https://api2.binance.com/api/v3/ticker/24hr?symbol=BTCUSDT'
+	) {
+		if ( document.querySelectorAll( '.js-external-data-embed' ).length === 0 && !insertFetch ) {
+			return;
+		}
+
+		const response = await fetch( externalUrl );
+		const data = await response.json();
+		const allItems = document.querySelectorAll( '.js-external-data-embed' );
+		const updateTime = new Date( data.closeTime );
+		const parsedData = '$' + Number( data.lastPrice ).toFixed( 2 ) + ' - ' + updateTime.toLocaleString();
+
+		allItems.forEach( item => {
+			item.classList.add( 'external-widget-bounce' );
+			item.textContent = parsedData;
+
+			setTimeout( () => item.classList.remove( 'external-widget-bounce' ), 2000 );
+		} );
+
+		return parsedData;
+	}
```

Comment:
You should not use `document.querySelectorAll()` since it scans the whole page (also other editor instances). Please use editor APIs to find model elements, sth like this:

```
for ( const { item } of model.createRangeIn( rootElement) ) { ... }
```

And trigger `editor.editing.reconvertItem( modelElement )` for all elements that should be refreshed in the view tree (but they were not altered in the model since the data is fetched from an external source and not stored in the model).

You should update the view elements in converters (they will be triggered by the above `reconvertItem()` call) and do not modify DOM in here.

Reply:
After when i implement all your tips - whole code became a much more cleaner, readable and as the intention was - to use CKEditor API superpowers.